pkg/yurthub/proxy/remote/remote.go

```diff
@@ -132,28 +132,34 @@ func (rp *RemoteProxy) modifyResponse(resp *http.Response) error {
 
 	// cache resp with storage interface
 	if resp.StatusCode >= http.StatusOK && resp.StatusCode <= http.StatusPartialContent {
```

Comment:
`respContentType` is not need by filterChain, so we can optimize this snippet as following.

```
	if resp.StatusCode >= http.StatusOK && resp.StatusCode <= http.StatusPartialContent {
		// filter response data
		if rp.filterChain != nil {
			size, filterRc, err := rp.filterChain.Filter(req, resp.Body, rp.stopCh)
			if err != nil {
				klog.Errorf("failed to filter response for %s, %v", util.ReqString(req), err)
				return err
			}
			resp.Body = filterRc
			if size > 0 {
				resp.ContentLength = int64(size)
				resp.Header.Set("Content-Length", fmt.Sprint(size))
			}
		}

		// cache resp with storage interface
		if rp.cacheMgr != nil && rp.cacheMgr.CanCacheFor(req) {
			reqContentType, _ := util.ReqContentTypeFrom(ctx)
			respContentType := resp.Header.Get("Content-Type")
			if len(respContentType) == 0 {
				respContentType = reqContentType
			}
			ctx = util.WithRespContentType(ctx, respContentType)
			req = req.WithContext(ctx)

			rc, prc := util.NewDualReadCloser(req, resp.Body, true)
			go func(req *http.Request, prc io.ReadCloser, stopCh <-chan struct{}) {
				err := rp.cacheMgr.CacheResponse(req, prc, stopCh)
				if err != nil && err != io.EOF && err != context.Canceled {
					klog.Errorf("%s response cache ended with error, %v", util.ReqString(req), err)
				}
			}(req, prc, rp.stopCh)

			resp.Body = rc
		}
	}
```

Reply:
It seems the `Filter` method of filters need `ContentType` to create the serializer:  
https://github.com/openyurtio/openyurt/blob/797c43d730a11a6cd9bd491f3d3838917e2a49cb/pkg/yurthub/filter/util/utils.go#L29-L38

Reply:
@rambohe-ch If we don't set  `respContentType` for filter, `CreateSerializer` will fail:
```
I0922 10:16:34.687792    7019 utils.go:34] CreateSerializer failed , info is :&{IsResourceRequest:true Path:/api/v1/services Verb:list APIPrefix:api APIGroup: APIVersion:v1 Namespace: Resource:services Subresource: Name: Parts:[services]}
E0922 10:16:34.687829    7019 filter.go:84] skip filter, failed to create serializer in masterServiceFilter
```

Reply:
@DrmagicE You are right, `filterChain` has used `responseContentType`, how about optimize as following:

```
	if resp.StatusCode >= http.StatusOK && resp.StatusCode <= http.StatusPartialContent {
		// prepare response content type
		reqContentType, _ := util.ReqContentTypeFrom(ctx)
		respContentType := resp.Header.Get("Content-Type")
		if len(respContentType) == 0 {
			respContentType = reqContentType
		}
		ctx = util.WithRespContentType(ctx, respContentType)
		req = req.WithContext(ctx)

		// filter response data
		if rp.filterChain != nil {
			size, filterRc, err := rp.filterChain.Filter(req, resp.Body, rp.stopCh)
			if err != nil {
				klog.Errorf("failed to filter response for %s, %v", util.ReqString(req), err)
				return err
			}
			resp.Body = filterRc
			if size > 0 {
				resp.ContentLength = int64(size)
				resp.Header.Set("Content-Length", fmt.Sprint(size))
			}
		}

		// cache resp with storage interface
		if rp.cacheMgr != nil && rp.cacheMgr.CanCacheFor(req) {
			rc, prc := util.NewDualReadCloser(req, resp.Body, true)
			go func(req *http.Request, prc io.ReadCloser, stopCh <-chan struct{}) {
				err := rp.cacheMgr.CacheResponse(req, prc, stopCh)
				if err != nil && err != io.EOF && err != context.Canceled {
					klog.Errorf("%s response cache ended with error, %v", util.ReqString(req), err)
				}
			}(req, prc, rp.stopCh)

			resp.Body = rc
		}
	}
```

Reply:
If both filter and cache manager are disabled, `respContentType` is not needed.

Reply:
If both filter and cache manager are disabled, `respContentType` is not needed.

Reply:
 I think it's not common to disable both filter and cache manager, and code as above is more readable. so maybe it's a better choice.

Reply:
@rambohe-ch agree, updated.